### DISCUSSION TITLE
Welcome to RgGen Discussions!

### DISCUSSION THREAD

**Discussion by taichi-ishitani** [rggen/rggen#48] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by nijie001** [comment]
I am new to RGGEN, any youtube training videos for RGGEN that I can learn from?

**Comment by short-stuff88** [comment]
I'm attempting to use rggen for a medical test ASIC as part of a university project.
As a result, I need to achieve a high level of quality and maintainability.  I need to demonstrate code and functional coverage.
I have a couple of thousand registers, many of which are reused in various RTL blocks.  The same bitfield type occurs many places.

There does not appear to be the following features in rggen:
1) a way to turn on UVM functional coverage, as a default, and then enable/disable  override on specific bitfields and registers
2) a way to declare a bitfield type (such as enum values, width, etc.) and then reuse it in multiple registers.  Thus one would say "create an instance of bitfield FOO at bit-offset 3" (since the bitfield type would already have its width predefined.
3) a way to declare that a bitfield is an enum type, not just a field of bits.
4) a way to declare a register-type and then reuse it (similar to the bit-field type).

My apologies if these features exist, and I've simply missed them - I've been reading the documentation and examples carefully.

**Reply by taichi-ishitani** [reply]
Hi @short-stuff88 

Thank you for interesting ReGen!

> a way to turn on UVM functional coverage, as a default, and then enable/disable override on specific bitfields and registers

Currently, no covergroups are generated. I'd like you to open an issue to describe your requirement.

>  a way to declare a bitfield type (such as enum values, width, etc.) and then reuse it in multiple registers. ?Thus one would say "create an instance of bitfield FOO at bit-offset 3" (since the bitfield type would already have its width predefined.

> a way to declare a register-type and then reuse it (similar to the bit-field type).

RgGen supports Ruby as an input foramt. By using Ruby format, you can define methods for such purpose like below.

```ruby
def bit_feild_foo(pos)
  bit_field {
    name: 'foo'
    bit_position lsb: pos, width: 4
    type :rw
    initial_value 0
  }
end

register_block {
  name 'block_0'
  byte_size 256
  
  regster {
    name 'reg_0'
    offset_address 0x0
    bit_feild_foo 0
  }

  regster {
    name 'reg_1'
    offset_address 0x4
    bit_feild_foo 4
  }
}
```

> a way to declare that a bitfield is an enum type, not just a field of bits.

I think `label` feature is suitable for this. https://github.com/rggen/rggen/issues/126
Sorry, I forgot documentation for this feature.
